File: dynamic_programming/Jump_Game.py
# https://leetcode.com/problems/jump-game-ii/

def jump(nums, memo, index=0):
    if index == len(nums) - 1:
        return 0
    result = 1001
    for i in range(1, nums[index]):
        if index + i < len(nums):
            result = min(result, 1+jump(nums, memo, index+i))
            print(result)
    memo[index] = result
    return result


arr = [2, 3, 1, 1, 4]
val = jump(arr, [0]*(len(arr)-1),0)
print(val)

File: greedy_algorithm/_go/jump_game.go
// https://leetcode.com/problems/jump-game/
package main

import (
	"fmt"
	"math"
)

func canJump(nums []int) bool {
	i, length, maxReach := 0, len(nums)-1, 0
	for i <= length && i <= maxReach {
		maxReach = int(math.Max(float64(maxReach), float64(i+nums[i])))
		i++
	}
	return i == len(nums)
}

func main() {
	//nums := []int{3, 2, 1, 0, 4}
	//nums := []int{2, 3, 1, 1, 4}
	nums := []int{2, 0}
	fmt.Println(canJump(nums))
}
